### PR TITLE
draw_image bi-linear interpolation, alpha palettes, and center_image

### PR DESCRIPTION
**Pull request by matsondawson**
There's an extra parameter on draw_image called hint.
It can take the values:
image.BILINEAR_INTERPOLATION which smooths images
image.IMAGE_CENTER which makes x,y coordinates relative to the center of the image.
e.g. hint=image.BILINEAR_INTERPOLATION | image.IMAGE_CENTER 

There's an extra parameter on draw_image called alpha_pallete
This works the same as color palette except it creates an alpha value based on the grayscale image.
This is very useful for combining thermal images to sensor when you want some areas to be clear with smoothed edges.

### CODE REVIEW

**Comment by kwagyeman** [comment]
Hi, what was wrong with what you had before? There are a lot of changes in one go. Can you explain what you are changing right now?

**Comment by kwagyeman** [comment]
That said, this fixes a request for thermal image blending someone had before.

**Comment by matsondawson** [comment]
> Hi, what was wrong with what you had before? There are a lot of changes in one go. Can you explain what you are changing right now in detail?

The current implementation uses nearest neighbor interpolation, this adds the ability to use linear smoothing between pixels, which means you can use lower resolution images for compositing thermal images.  This change is actually for the other guy whom I'm working with.

The old code path is still used if you don't pass the interpolation hint, so if it were buggy somehow people can just not use the hint.

I've attached a comparison image.

filtering

**Comment by kwagyeman** [comment]
Sorry, I just looked at this after an all day coding session and was like, man, this is a lot of changes.

File: src/omv/img/imlib.h [hunk 1]
```diff
@@ -1141,6 +1141,12 @@ typedef struct find_barcodes_list_lnk_data {
     int quality;
 } find_barcodes_list_lnk_data_t;
 
+typedef enum image_hint {
```
Comment (kwagyeman) [inline src/omv/img/imlib.h:1144]: Can you rename these with a PREFIX. Like IMAGE_HINT_BILINEAR, IMAGE_HINT_CENTER.

File: src/omv/img/imlib.h [hunk 2]
```diff
@@ -1141,6 +1141,12 @@ typedef struct find_barcodes_list_lnk_data {
...
+    INTERPOLATE_BILINEAR = 1,
+    IMAGE_CENTER = 128
+} image_hint_type;
```
Comment (kwagyeman) [inline src/omv/img/imlib.h:1147]: image_hint_t

File: src/omv/py/qstrdefsomv.h [hunk 3]
```diff
@@ -551,6 +552,9 @@ Q(draw_image)
 // duplicate Q(y_scale)
 Q(alpha)
 // duplicate Q(mask)
+Q(hint)
+Q(INTERPOLATE_BILINEAR)
+Q(IMAGE_CENTER)
```
Comment (kwagyeman) [inline src/omv/py/qstrdefsomv.h:557]: Can you rename these with a PREFIX. Like IMAGE_HINT_BILINEAR, IMAGE_HINT_CENTER.

File: src/omv/py/py_image.c [hunk 4]
```diff
@@ -7594,6 +7625,8 @@ static const mp_rom_map_elem_t globals_dict_table[] = {
     {MP_ROM_QSTR(MP_QSTR_CODE93),              MP_ROM_INT(BARCODE_CODE93)},
     {MP_ROM_QSTR(MP_QSTR_CODE128),             MP_ROM_INT(BARCODE_CODE128)},
 #endif
+    {MP_ROM_QSTR(MP_QSTR_INTERPOLATE_BILINEAR),MP_ROM_INT(INTERPOLATE_BILINEAR)},
+    {MP_ROM_QSTR(MP_QSTR_IMAGE_CENTER),        MP_ROM_INT(IMAGE_CENTER)},
```
Comment (kwagyeman) [inline src/omv/py/py_image.c:7629]: Can you rename these with a PREFIX. Like IMAGE_HINT_BILINEAR, IMAGE_HINT_CENTER.

File: src/omv/img/imlib.h [hunk 5]
```diff
@@ -259,7 +259,7 @@ extern const uint8_t g826_table[256];
     __typeof__ (r5) _r5 = (r5); \
     __typeof__ (g6) _g6 = (g6); \
     __typeof__ (b5) _b5 = (b5); \
-    (_r5 << 3) | (_g6 >> 3) | ((_g6 & 0x7) << 13) | (_b5 << 8); \
+    __REV16((_r5 << 11) | (_g6 << 5) | _b5); \
```
Comment (kwagyeman) [inline src/omv/img/imlib.h:262]: Cool.

File: src/omv/img/draw.c [hunk 6]
```diff
@@ -488,21 +488,217 @@ static int safe_map_pixel(int dst_bpp, int src_bpp, int pixel)
...
+    for (int i = 0, x = other_x_start; x < other_x_end; x++, i++) {
+        float other_x_float = x * over_xscale;
+        uint32_t other_x = fast_floorf(other_x_float);
+        uint32_t weight_x = fast_floorf((other_x_float - other_x) * alpha);
+
+        bool mask1 = true, mask2 = true;
+        if (mask_row_ptr) {
+            mask1 = pixel_to_binary(mask_bpp, imlib_get_pixel_fast(mask_bpp, mask_row_ptr, other_x));
+            mask2 = pixel_to_binary(mask_bpp, imlib_get_pixel_fast(mask_bpp, mask_row_ptr, other_x + 1));
+        }
```
Comment (kwagyeman) [inline src/omv/img/draw.c:594]: space

File: src/omv/img/draw.c [hunk 7]
```diff
@@ -488,21 +488,217 @@ static int safe_map_pixel(int dst_bpp, int src_bpp, int pixel)
...
+ */
+static void int_generate_cache_line_grayscale(uint16_t *cache_line, int alpha, uint8_t *other_row_ptr, int other_bpp, void *mask_row_ptr, int mask_bpp, int other_x_start, int other_x_end, float over_xscale)
+{
+    // generate line
+    for (int i = 0, x = other_x_start; x < other_x_end; x++, i++) {
+        float other_x_float = x * over_xscale;
+        uint32_t other_x = fast_floorf(other_x_float);
+        uint32_t weight_x = fast_floorf((other_x_float - other_x) * alpha);
+
+        bool mask1 = true, mask2 = true;
```
Comment (kwagyeman) [inline src/omv/img/draw.c:590]: move up and space afterwards to match RGB888 code below.

File: src/omv/img/draw.c [hunk 8]
```diff
@@ -551,109 +757,251 @@ void imlib_draw_image(image_t *img, image_t *other, int x_off, int y_off, float
...
-                        IMAGE_PUT_GRAYSCALE_PIXEL_FAST(img_row_ptr, x, result_pixel);
+                        // Store pixel
+                        IMAGE_PUT_GRAYSCALE_PIXEL_FAST(img_row_ptr, x, pixel);
+                    }
+                }
+                // De-allocate cache lines
+                fb_free(cache_line_1);
+                fb_free(cache_line_2);
+            }
+            else {
```
Comment (kwagyeman) [inline src/omv/img/draw.c:840]: } else {

File: src/omv/img/draw.c [hunk 9]
```diff
@@ -551,109 +757,251 @@ void imlib_draw_image(image_t *img, image_t *other, int x_off, int y_off, float
...
+                        // Store pixel
                         IMAGE_PUT_RGB565_PIXEL_FAST(img_row_ptr, x, result_pixel);
                     }
                 }
+                
+                // De-allocate cache lines
+                fb_free(cache_line_1);
+                fb_free(cache_line_2);
+            }
+            else {
```
Comment (kwagyeman) [inline src/omv/img/draw.c:963]: } else {

**Comment by kwagyeman** [comment]
Supporting the negative scale stuff makes this code too complex. It's kinda at the point where you need to be too smart to debug it.

**Comment by matsondawson** [comment]
> Supporting the negative scale stuff makes this code too complex. It's kinda at the point where you need to be too smart to debug it.

I'll look into it, my spider sense says there may be something up with it.

**Comment by matsondawson** [comment]
I'm thinking of removing the negative scales.  I think if you want negative stuff you can use a flip command or something.

**Comment by kwagyeman** [comment]
An issue failed with travis pulling data. Need to kick it again.

**Comment by kwagyeman** [comment]
Can merge this now.

**Comment by iabdalkader** [comment]
I'll review and merge asap. sorry was busy with I2C and an other project.

**Comment by iabdalkader** [comment]
> > Hi, what was wrong with what you had before? There are a lot of changes in one go. Can you explain what you are changing right now in detail?
> 
> The current implementation uses nearest neighbor interpolation, this adds the ability to use linear smoothing between pixels, which means you can use lower resolution images for compositing thermal images. This change is actually for the other guy whom I'm working with.
> 
> The old code path is still used if you don't pass the interpolation hint, so if it were buggy somehow people can just not use the hint.
> 
> Interestingly if the image is scaled at least 2x then the bi-linear filtering code draws faster than the nearest neighbor code.
> 
> filtering

Thank you for the PR. Do you mind if we tweet this image to show the improved blending ? And if you'd like to share more details about what you're doing please let us know, we could write a blog post about it.

**Pull request merged**